appsec/src/extension/user_tracking.c

```diff
@@ -61,6 +67,88 @@ static PHP_FUNCTION(set_user_wrapper)
     _ddtrace_set_user(INTERNAL_FUNCTION_PARAM_PASSTHRU);
 }
 
+static void _emit_user_event(void)
+{
+    dd_tags_set_user_event_triggered();
+    dd_trace_emit_asm_event();
+    dd_tags_set_sampling_priority();
+}
+
+static PHP_FUNCTION(v2_track_user_login_success_wrapper)
+{
+    _ddtrace_v2_track_user_login_success(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    dd_telemetry_add_sdk_event(LSTRARG("login_success"));
+
+    zend_string *login;
+    zval *user = NULL;
+    zend_array *metadata = NULL;
+    zend_string *user_id = NULL;
+    if (!DDAPPSEC_G(active) && UNEXPECTED(!get_global_DD_APPSEC_TESTING())) {
+        return;
+    }
+    if (zend_parse_parameters(
+            ZEND_NUM_ARGS(), "S|zh", &login, &user, &metadata) == FAILURE) {
+        return;
+    }
+    if (_ddtrace_v2_track_user_login_success == NULL) {
+        mlog(dd_log_debug, "Invalid DDTrace\\track_user_login_success, "
+                           "this shouldn't happen");
+        return;
+    }
+
+    if (user != NULL && Z_TYPE_P(user) == IS_STRING) {
+        user_id = Z_STR_P(user);
+    } else if (user != NULL && Z_TYPE_P(user) == IS_ARRAY) {
+        zval *user_id_zv = zend_hash_str_find(Z_ARR_P(user), ZEND_STRL("id"));
+        if (user_id_zv == NULL) {
+            mlog(dd_log_warning,
+                "Id not found in user object in "
+                "datadog\\appsec\\v2\\track_user_login_success");
+            return;
+        }
+        if (Z_TYPE_P(user_id_zv) != IS_STRING) {
+            mlog(dd_log_warning,
+                "Unexpected id type in "
+                "datadog\\appsec\\v2\\track_user_login_success");
+            return;
+        }
+        user_id = Z_STR_P(user_id_zv);
+    }
+
+    _emit_user_event();
+    dd_find_and_apply_verdict_for_user(
+        user_id, login, user_event_login_success);
+}
+
+static PHP_FUNCTION(v2_track_user_login_failure_wrapper)
+{
+    _ddtrace_v2_track_user_login_failure(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    dd_telemetry_add_sdk_event(LSTRARG("login_failure"));
+
+    zend_string *login = NULL;
+    zend_bool exists;
+    zend_array *metadata = NULL;
+
+    if (!DDAPPSEC_G(active) && UNEXPECTED(!get_global_DD_APPSEC_TESTING())) {
+        return;
+    }
+
+    if (zend_parse_parameters(
+            ZEND_NUM_ARGS(), "Sb|h", &login, &exists, &metadata) == FAILURE) {
+        return;
+    }
+
+    if (_ddtrace_v2_track_user_login_failure == NULL) {
+        mlog(dd_log_debug, "Invalid DDTrace\\track_user_login_failure, "
+                           "this shouldn't happen");
+        return;
+    }
+
+    _emit_user_event();
```

Comment:
This should happen right after calling `_ddtrace_v2_track_user_login_failure`, because you have exit conditions before this point but the sampling priority must be set for SDK events.